File: food-ordering-system/go-without-multithreading/strategy/max_capacity_strategy.go
package strategy
import "app/food-ordering-system/go-without-multithreading/entities"

type MaxCapacityStrategy struct{}

func (s *MaxCapacityStrategy) SelectRestaurant(eligibleRestaurants []*entities.Restaurant, requestedItems map[string]int) *entities.Restaurant {
	var bestRestaurant *entities.Restaurant
	maxRemaining := -1

	for _, r := range eligibleRestaurants {
		remaining := r.MaxOrders - r.CurrentOrders
		if remaining > maxRemaining {
			maxRemaining = remaining
			bestRestaurant = r
		}
	}
	return bestRestaurant
}

File: foodOrderingSystem/java-without-multithreading/strategy/MaxCapacityStrategy.java
package strategy;
import entities.Restaurant;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class MaxCapacityStrategy implements RestaurantSelectionStrategy {
    @Override
    public Optional<Restaurant> selectRestaurant(List<Restaurant> eligibleRestaurants, Map<String, Integer> requestedItems) {
        Restaurant bestRestaurant = null;
        int maxRemaining = -1;

        for (Restaurant r : eligibleRestaurants) {
            int remaining = r.getMaxOrders() - r.getCurrentOrders();
            if (remaining > maxRemaining) {
                maxRemaining = remaining;
                bestRestaurant = r;
            }
        }
        return Optional.ofNullable(bestRestaurant);
    }
}
